<a href="https://colab.research.google.com/github/ElBaron-27/UACJ_Ruben/blob/main/Sesion12_Evaluacion_Datos_Categoricos_Ruben_Uribe.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Programación para Analítica Descriptiva y Predictiva**
**Maestría en Inteligencia Artificial y Analítica de Datos**

# Sesión 12: Evaluación — Limpieza y Transformación de Datos Categóricos

**Entrega individual**

- **Nombre**: Ruben Alejandro Uribe Rodriguez
- **Matrícula** 271774

Esta evaluación aplica los tres temas de la Sesión 12 (errores tipográficos y valores inconsistentes, alta cardinalidad, tipos incorrectos) a un dataset que no se trabajó en clase: **Telco Customer Churn**.


No hay una única respuesta correcta en varias de las actividades — lo que se evalúa es que la conclusión esté respaldada por el código que la sustenta, no solo la conclusión en sí.

## Preparación

In [1]:
import pandas as pd
import numpy as np

url = 'https://raw.githubusercontent.com/treselle-systems/customer_churn_analysis/master/WA_Fn-UseC_-Telco-Customer-Churn.csv'
df = pd.read_csv(url)
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


---
## Actividad 1 — Formato y valores inconsistentes (15 pts)

Revisa **todas** las columnas categóricas del dataset (no elijas solo una) en busca de variantes de formato (mayúsculas, espacios) que deberían normalizarse.

In [3]:
# Revisamos los valores únicos de cada columna para identificar
# posibles diferencias de formato, espacios o categorías inconsistentes.

for col in df.columns:
  print(f"Valores unicos en {col}: {df[col].unique()}")

# Revisamos también la frecuencia de cada valor para detectar
# categorías atípicas o variantes poco frecuentes.

for col in df.columns:
  print(f"Frecuencia de Valores únicos en {col}: {df[col].value_counts()}")

Valores unicos en customerID: ['7590-VHVEG' '5575-GNVDE' '3668-QPYBK' ... '4801-JZAZL' '8361-LTMKD'
 '3186-AJIEK']
Valores unicos en gender: ['Female' 'Male']
Valores unicos en SeniorCitizen: [0 1]
Valores unicos en Partner: ['Yes' 'No']
Valores unicos en Dependents: ['No' 'Yes']
Valores unicos en tenure: [ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30 47 72 17 27
  5 46 11 70 63 43 15 60 18 66  9  3 31 50 64 56  7 42 35 48 29 65 38 68
 32 55 37 36 41  6  4 33 67 23 57 61 14 20 53 40 59 24 44 19 54 51 26  0
 39]
Valores unicos en PhoneService: ['No' 'Yes']
Valores unicos en MultipleLines: ['No phone service' 'No' 'Yes']
Valores unicos en InternetService: ['DSL' 'Fiber optic' 'No']
Valores unicos en OnlineSecurity: ['No' 'Yes' 'No internet service']
Valores unicos en OnlineBackup: ['Yes' 'No' 'No internet service']
Valores unicos en DeviceProtection: ['No' 'Yes' 'No internet service']
Valores unicos en TechSupport: ['No' 'Yes' 'No internet service']
Valores unicos en Stream

**1.2 — Conclusión (responde aquí en Markdown):**

¿Encontraste alguna columna con inconsistencias de formato? Si sí, ¿cuál y qué código usarías para corregirla? Si no encontraste ninguna, dilo explícitamente — es una conclusión válida siempre que esté respaldada por lo que revisaste en 1.1.

_Tu respuesta:_No, no encontré ninguna columna con inconsistencias de formato de mayúsculas/minúsculas o espacios en blanco. Todas las columnas categóricas muestran categorías sin discrepancias entre mayus/minus y sin espacios sobrantes. Por lo tanto, concluyo que no se requiere código de normalización de texto para este dataset.

---
## Actividad 2 — Valores inválidos (20 pts)

Varias columnas de este dataset (`OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`, `StreamingMovies`) tienen un tercer valor además de `'Yes'`/`'No'`: `'No internet service'`. De forma similar, `MultipleLines` tiene `'No phone service'`.

In [4]:
df['OnlineSecurity'].value_counts()

,count
OnlineSecurity,
No,3498
Yes,2019
No internet service,1526


In [5]:
# 2.1 — Verifica: ¿las filas con 'No internet service' en OnlineSecurity
# coinciden con las filas donde InternetService == 'No'?
# (pista: cruza ambas columnas con pd.crosstab o filtrando)
pd.crosstab(df['InternetService'], df['OnlineSecurity'])

# Usamos una tabla cruzada para comprobar si ambas categorías coinciden
# y decidir si "No internet service" representa un error o información válida.


OnlineSecurity,No,No internet service,Yes
InternetService,,,
DSL,1241,0,1180
Fiber optic,2257,0,839
No,0,1526,0


**2.2 — Conclusión (responde aquí en Markdown):**

¿`'No internet service'` es un valor inválido (como `Absurd`/`YOLO` en la sesión de clase) o es una categoría legítima? Justifica tu respuesta con lo que verificaste en 2.1. ¿Tomarías alguna acción sobre esta columna, o la dejarías tal cual?

_Tu respuesta:_ Es una categoría legítima, sin embargo, yo la removería ya que es redundante. Como se puede observar la categoría 'No internet service' aparece exclusivamente en los registros donde InternetService == 'No' (1526 de 1526 casos), sin ninguna intersección con las categorías DSL o Fiber optic. Esto confirma que esa categoría no aporta información adicional independiente, sino que es una redundancia derivada directamente de la variable InternetService.En mi caso yo la simplificaria colapsandola en la categoria 'No'

---
## Actividad 3 — Alta cardinalidad (25 pts)

In [6]:
# 3.1 — Calcula .nunique() para TODAS las columnas del dataset (no solo las categóricas)
# y la razón (valores únicos / total de filas) para cada una.

unicos = df.nunique()
razon = unicos / len(df)
resumen = pd.DataFrame({
    'valores_unicos': unicos,
    'razon_unicidad': razon
}).sort_values('razon_unicidad', ascending=False)

print(resumen)

                  valores_unicos  razon_unicidad
customerID                  7043        1.000000
TotalCharges                6531        0.927304
MonthlyCharges              1585        0.225046
tenure                        73        0.010365
PaymentMethod                  4        0.000568
StreamingMovies                3        0.000426
TechSupport                    3        0.000426
OnlineBackup                   3        0.000426
StreamingTV                    3        0.000426
DeviceProtection               3        0.000426
MultipleLines                  3        0.000426
InternetService                3        0.000426
OnlineSecurity                 3        0.000426
Contract                       3        0.000426
Partner                        2        0.000284
SeniorCitizen                  2        0.000284
gender                         2        0.000284
Dependents                     2        0.000284
PhoneService                   2        0.000284
PaperlessBilling    

**3.2 — Conclusión (responde aquí en Markdown):**

¿Qué columna(s) tienen alta cardinalidad? Para la columna con mayor cardinalidad: ¿por qué nunca deberías usarla como variable predictora en un modelo, incluso si la codificaras? (relaciona tu respuesta con lo discutido en clase sobre identificadores únicos)

_Tu respuesta:_"customerID", "TotalCharges", "MonthlyCharges", sin embargo, para estos últimos 2, tecnicamente no serían problemáticos dado que son variables numéricas continuas. En el caso de "customerID", al ser un identificador único no aporta para el modelo predictivo y se tiene que eliminar antes de seguir trabajando con el dataset.

**3.3 — Agrupación "Top 10 + Otros"**

En clase agrupaste `country` de Netflix Titles en sus 10 categorías más frecuentes + `'Otros'`, reduciendo su cardinalidad. Aplica la misma técnica aquí sobre la columna de mayor cardinalidad que identificaste en 3.1 (pista: `.value_counts().head(10)`, luego `.where()` + `.isin()`, igual que en el notebook de clase).

In [7]:
# Aplica el agrupamiento Top 10 + Otros sobre la columna de mayor cardinalidad

# Identificamos los 10 customerID más frecuentes para aplicar
top_10_customer_ids = df['customerID'].value_counts().head(10).index


# Conservamos esos 10 valores y agrupamos todos los demás como "Otros".
# Posteriormente evaluaremos si esta transformación tiene sentido para un ID.
df['customerID_agrupados'] = df['customerID'].where(df['customerID'].isin(top_10_customer_ids), 'Otros')

print(df['customerID_agrupados'].value_counts())

customerID_agrupados
Otros         7033
7590-VHVEG       1
5575-GNVDE       1
9837-FWLCH       1
1699-HPSBG       1
7203-OYKCT       1
1035-IPQPU       1
7398-LXGYX       1
2823-LKABH       1
8775-CEBBJ       1
3186-AJIEK       1
Name: count, dtype: int64


**3.4 — Conclusión (responde aquí en Markdown):**

Después de agrupar, ¿la columna resultante te parece útil para un modelo? Compara este caso con el de `country` en el notebook de clase: ¿por qué agrupar en "Top 10 + Otros" funciona bien para una variable como `country`, pero no resuelve el problema real de la columna que agrupaste aquí?

_Tu respuesta:_Porque en el caso de 'country' se trataba de columnas con muchas categorías en donde pocos países concentraban el 80% de los datos y el resto son casos raros, por lo que era más práctico concentrarlos en una categoría diferente al ser el porcentaje menor de datos. En este caso me devolvió los 10 IDs de forma arbitraria y  no hay "los 10 más frecuentes" de verdad, porque no hay frecuencias que comparar ya que todos son igual de frecuentes (una vez). Hacer este tipo de columna no es útil para este tipo de modelos

---
## Actividad 4 — Tipos de dato (30 pts)

In [8]:
df['TotalCharges'].dtype

dtype('O')

**4.1 — Investiga (responde en Markdown):**

`TotalCharges` contiene valores numéricos (montos en dólares), pero pandas la cargó como `object`, no como `float`. Investiga por qué — revisa si hay algún valor que no se vea como un número normal.

_Tu respuesta:_ Aunque se observan que la mayoría de los valores son numéricos se encontraron 11 registros que contienen un espacio en blanco (' ') en lugar de un número. Algo importantes es que al revisar estos registros, también se observa que corresponden a clientes con tenure = 0, por lo que probablemente todavía no cuentan con cargos totales acumulados. En conclusión, el problema no se encuentra en los valores numéricos de la columna, sino en esos valores vacíos almacenados como texto.

In [9]:
# 4.2 — Corrige el tipo de TotalCharges.
# Pista: pd.to_numeric() con el parámetro errors= te puede ayudar a identificar
# o manejar los valores problemáticos que encontraste en 4.1.

# Identificamos los valores vacíos de TotalCharges
totalcharges_vacios = df['TotalCharges'].str.strip() == ''

print(f"Valores vacíos encontrados: {totalcharges_vacios.sum()}")

# Los registros vacíos corresponden a clientes con tenure = 0,
# por lo que sustituimos TotalCharges por 0 sin eliminar las filas
df.loc[totalcharges_vacios, 'TotalCharges'] = '0'

# Convertimos finalmente la columna a tipo numérico
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='raise')

# Verificamos el resultado
print(df['TotalCharges'].dtype)
print(f"Valores nulos: {df['TotalCharges'].isna().sum()}")

Valores vacíos encontrados: 11
float64
Valores nulos: 0


In [10]:
# 4.3 — Convierte a category las columnas categóricas que, según lo que calculaste
# en la Actividad 3, tengan cardinalidad baja y valores fijos.
# Verifica con .dtypes que el cambio se aplicó correctamente.

columnas_category = [
    'gender',
    'SeniorCitizen',
    'Partner',
    'Dependents',
    'PhoneService',
    'MultipleLines',
    'InternetService',
    'OnlineSecurity',
    'OnlineBackup',
    'DeviceProtection',
    'TechSupport',
    'StreamingTV',
    'StreamingMovies',
    'Contract',
    'PaperlessBilling',
    'PaymentMethod',
    'Churn'
]


for col in columnas_category:
    df[col] = df[col].astype('category')

print(df[columnas_category].dtypes)


gender              category
SeniorCitizen       category
Partner             category
Dependents          category
PhoneService        category
MultipleLines       category
InternetService     category
OnlineSecurity      category
OnlineBackup        category
DeviceProtection    category
TechSupport         category
StreamingTV         category
StreamingMovies     category
Contract            category
PaperlessBilling    category
PaymentMethod       category
Churn               category
dtype: object


---
## Reflexión final (10 pts)

Con base en las 4 actividades anteriores, responde:

1. De las alertas que detectaste (formato, valores inválidos, cardinalidad, tipos), ¿cuál te pareció más fácil de decidir y cuál más difícil? ¿Por qué?
2. Si tuvieras que entregar este dataset ya "perfilado" a un compañero para que construya un modelo predictivo, ¿qué le dirías sobre `customerID` y sobre `TotalCharges`?

_Tu respuesta:_ 1. Más fácil fue lo de valores inválidos ya que inmediatamente pude observar que este dataset no tenía ese problema. y más difícil fue cardinalidad, ya que si bien la columna de ID era obvia, la parte de 'tenure' me causó un poco de duda. E igual en la pregunta de tipos me tardé un rato en encontrar la relación con 'tenure'. Con respecto a la pregunta 2. en primer lugar recomendaría omitir la columna 'customerID' ya que es intrascendente para nuestro modelo predictivo y en cuanto a 'totalcharges' le explicaría que originalmente fue cargada como object debido a 11 valores en blanco correspondientes a clientes con tenure = 0 sin embargo estos valores fueron tratados como 0 para conservar los registros y posteriormente la columna fue convertida a un tipo numérico. Así TotalCharges puede mantenerse como una variable cuantitativa que será muy útil para el modelo.

---
## Rúbrica de evaluación

| Actividad | Puntos | Criterio |
|---|---|---|
| 1. Formato y valores inconsistentes | 15 | Revisó todas las columnas categóricas (no solo una); código comentado y conclusión (1.2) respaldada por lo que se observó, no solo afirmada |
| 2. Valores inválidos | 20 | Verificó la relación entre columnas antes de concluir; la conclusión (2.2) justifica con evidencia, no solo con intuición |
| 3. Alta cardinalidad | 25 | Calcula `.nunique()` para todas las columnas; aplica correctamente el agrupamiento Top 10 + Otros; la conclusión (3.4) explica por qué agrupar no resuelve el problema de un identificador único |
| 4. Tipos de dato | 30 | Identifica la causa raíz del `dtype` incorrecto (4.1); corrige `TotalCharges` sin perder información; conversión a `category` justificada por cardinalidad, no aplicada al azar |
| Reflexión final | 10 | Conecta las 4 actividades entre sí; no es una respuesta genérica o intercambiable con cualquier dataset |
| **Total** | **100** | |

**Nota sobre las conclusiones:** cada actividad tiene su propia pregunta de conclusión (1.2, 2.2, 3.4, 4.1) — esas respuestas se califican como parte de la actividad correspondiente, no solo la Reflexión final. Comenta tu código donde tomes una decisión (por ejemplo, por qué elegiste cierto umbral o cierta corrección) — el comentario también cuenta dentro del puntaje de cada actividad.